segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentContainer.java

```diff
@@ -25,4 +26,16 @@
      * @return A Collection with the SegmentProperties for these segments.
      */
     Collection<SegmentProperties> getActiveSegments();
+
+    /**
+     * Returns a DirectSegmentAccess object that can be used for operating on a particular StreamSegment directly. The
+     * result of this call should only be used for short periods of time (i.e., while processing a single request) and
+     * not be cached for long-term access.
+     *
+     * @param streamSegmentName The name of the StreamSegment to get for.
+     * @param timeout           Timeout for the operation.
+     * @return A CompletableFuture that, when completed normally, will contain the desired result. If the operation
+     * failed, the future will be failed with the causing exception.
+     */
+    CompletableFuture<DirectSegmentAccess> forSegment(String streamSegmentName, Duration timeout);
```

Comment:
While I understand the performance benefits of a proper use of a `DirectSegmentAccess` instance, I'm not sure about the potential consequences of an (unintentional) improper use of it. Could we face any extra issues of operating on segments via a `DirectSegmentAccess` instance compared with `StreamSegmentContainer` (e.g., parallel updates on a segment, stale state)? Would it make sense to create a test verifying the extent of some "improper usage" cases of `DirectSegmentAccess`?

Reply:
The Javadoc is clear that instances of `DirectSegmentAccess` are short-lived and should not be cached for future access. Any such attempt will either succeed (if the segment is still active) or fail (saying segment is not registered). 

The benefits outweigh these problems: it allows one to pass a `DirectSegmentAccess` instance to one or more methods which can then execute on that. The alternative would have been to pass "Segment Name (String)" and "Container" to all such methods, which would make the API look more complicated than necessary and expose methods that we do not want (`SegmentContainer` exposes container-specific operations that have nothing to do with segments).

Another benefit is ease of mocking in unit tests. It's much easier to mock a `DirectSegmentAccess` object than having to mock an entire `SegmentContainer` (with all the extra bells and whistles). This will be very apparent in one of my (not-yet-filed) PRs relating to Tables.

Reply:
Plus this is only exposed in the `SegmentContainer` API. Nobody outside of the `segmentstore:server` gradle project has access to this API. It's purely internal and it is meant to be used by Segment Container Extensions (see PR #2861). At that level, you can already do a lot of bad things with those APIs if not properly used.